smac/optimizer/local_bo/abstract_subspace.py

```diff
@@ -0,0 +1,474 @@
+from smac.optimizer.acquisition import EI, TS
+from smac.optimizer.ei_optimization import AcquisitionFunctionMaximizer
+
+
+class AbstractSubspace(ABC):
+    def __init__(self,
+                 config_space: ConfigurationSpace,
+                 bounds: typing.List[typing.Tuple[float, float]],
+                 hps_types: typing.List[int],
+                 bounds_ss_cont: typing.Optional[np.ndarray] = None,
+                 bounds_ss_cat: typing.Optional[typing.List[typing.Tuple]] = None,
+                 model_local: AbstractEPM = PartialSparseGaussianProcess,
+                 model_local_kwargs: typing.Optional[typing.Dict] = None,
+                 acq_func_local: AbstractAcquisitionFunction = TS,
+                 acq_func_local_kwargs: typing.Optional[typing.Dict] = None,
+                 rng: typing.Optional[np.random.RandomState] = None,
+                 initial_data: typing.Optional[typing.Tuple[np.ndarray, np.ndarray]] = None,
+                 activate_dims: typing.Optional[np.ndarray] = None,
+                 incumbent_array: typing.Optional[np.ndarray] = None,
+                 ):
+        """
+        A subspace that is designed for local Bayesian Optimization
+        Parameters
+        ----------
+        config_space: ConfigurationSpace
+            raw Configuration space
+        bounds: typing.List[typing.Tuple[float, float]]
+            raw bounds of the Configuration space
+        hps_types: typing.List[int],
+            types of the hyperparameters
+        bounds_ss_cont: np.ndarray(D_cont, 2)
+            subspaces bounds of continuous hyperparameters, its length is the number of continuous hyperparameters
+        bounds_ss_cat: typing.List[typing.Tuple]
+            subspaces bounds of categorical hyperparameters, its length is the number of categorical hyperparameters
+        rng: np.random.RandomState
+            random state
+        model_local: ~smac.epm.base_epm.AbstractEPM
+            model in subspace
+        model_local_kwargs: typing.Optional[typing.Dict]
+            argument for subspace model
+        acq_func_local: ~smac.optimizer.ei_optimization.AbstractAcquisitionFunction
+            local acquisition function
+        acq_func_local_kwargs: typing.Optional[typing.Dict]
+            argument for acquisiiton fucntion
+        activate_dims: typing.Optional[np.ndarray]
+            activate dimensions in the subspace, if it is None, we preserve all the dimensions
+        incumbent_array: typing.Optional[np.ndarray]
+            incumbent array, used when activate_dims has less dimension and this value is used to complementary the
+            resulted configurations
+        """
+        self.logger = logging.getLogger(
+            self.__module__ + "." + self.__class__.__name__)
+        self.cs_global = config_space
+        if rng is None:
+            self.rng = np.random.RandomState(1)
+        else:
+            self.rng = np.random.RandomState(rng.randint(0, 2 ** 20))
```

Comment:
I would have expected that if `rng` is not None it would already be a numpy RandomState.
At least according to the above docstring.

Reply:
To avoid using the same rng from outside as this could probably be a reference?